File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/ExecutingSqlQueries.PetaPoco.cs
using System;
using System.Linq;
using System.Threading;
using PetaPoco;

class Program
{
    public void IExecuteMethods(IExecute execute, string query, string param)
    {
        execute.Execute(query);                                                                            // Compliant
        execute.Execute(query + param);                                                                    // Noncompliant

        execute.ExecuteScalar<Entity>(query);                                                              // Compliant
        execute.ExecuteScalar<Entity>(query + param);                                                      // Noncompliant
    }

    public void IDatabaseMethods(IDatabase database, string query, int param, int otherParam)
    {
        database.Execute(query);                                                                           // Compliant
        database.Execute(query + param);                                                                   // Noncompliant

        database.ExecuteScalar<Entity>(query);                                                             // Compliant
        database.ExecuteScalar<Entity>(query + param);                                                     // Noncompliant

        database.Query<Entity>(query);                                                                     // Compliant
        database.Query<Entity>(query + param);                                                             // Noncompliant
        database.Query<Entity>(query, param + otherParam);                                                 // Compliant
        database.Query<Entity>(new[] { typeof(Entity) }, null, query + param);                             // Noncompliant
        database.Query<Entity, Entity>(query + param);                                                     // Noncompliant
        database.Query<Entity, Entity, Entity>(query + param);                                             // Noncompliant
        database.Query<Entity, Entity, Entity, Entity>(query + param);                                     // Noncompliant
        database.Query<Entity, Entity, Entity, Entity, Entity>(query + param);                             // Noncompliant

        database.Fetch<Entity>(query);                                                                     // Compliant
        database.Fetch<Entity>(query + param);                                                             // Noncompliant
        database.Fetch<Entity>(2, 10, query + param);                                                      // Noncompliant
        database.Fetch<Entity>(param + otherParam, 10, query);                                             // Compliant
        database.Page<Entity>(0, 0, query + param, new object[0], query + param, new object[0]);
        //                          ^^^^^^^^^^^^^
        //                                                        ^^^^^^^^^^^^^                               @-1

        // All IQuery members
        database.Exists<Entity>(query + param);                                                            // Noncompliant
        database.Fetch<Entity, Entity>(query + param);                                                     // Noncompliant
        database.First<Entity>(query + param);                                                             // Noncompliant
        database.FirstOrDefault<Entity>(query + param);                                                    // Noncompliant
        database.Page<Entity>(0, 0, query + param);                                                        // Noncompliant (Parameter "sql")
        database.Page<Entity>(0, 0, query + param, Array.Empty<object>(), "", Array.Empty<object>());      // Noncompliant (Parameter "sqlCount")
        database.Page<Entity>(0, 0, "", Array.Empty<object>(), query + param, Array.Empty<object>());      // Noncompliant (Parameter "sqlPage")
        database.Query<Entity, Entity>(query + param);                                                     // Noncompliant
        database.QueryMultiple(query + param);                                                             // Noncompliant
        database.SkipTake<Entity>(0, 0, query + param);                                                    // Noncompliant
        database.Single<Entity>(query + param);                                                            // Noncompliant
        database.SingleOrDefault<Entity>(query + param);                                                   // Noncompliant

        // All IQueryAsync members
        database.ExistsAsync<Entity>(query + param);                                                       // Noncompliant
        database.FetchAsync<Entity>(query + param);                                                        // Noncompliant
        database.FirstAsync<Entity>(query + param);                                                        // Noncompliant
        database.FirstOrDefaultAsync<Entity>(query + param);                                               // Noncompliant
        database.PageAsync<Entity>(0, 0, query + param);                                                   // Noncompliant (Parameter "sql")
        database.PageAsync<Entity>(0, 0, query + param, Array.Empty<object>(), "", Array.Empty<object>()); // Noncompliant (Parameter "sqlCount")
        database.PageAsync<Entity>(0, 0, "", Array.Empty<object>(), query + param, Array.Empty<object>()); // Noncompliant (Parameter "sqlPage")
        database.QueryAsync<Entity>(query + param);                                                        // Noncompliant
        database.SkipTakeAsync<Entity>(0, 0, query + param);                                               // Noncompliant
        database.SingleAsync<Entity>(query + param);                                                       // Noncompliant
        database.SingleOrDefaultAsync<Entity>(query + param);                                              // Noncompliant

        // All IExecute and IExecuteAsync
        database.Execute(query + param);                                                                   // Noncompliant
        database.ExecuteAsync(CancellationToken.None, query + param);                                      // Noncompliant
        database.ExecuteScalar<Entity>(query + param);                                                     // Noncompliant
        database.ExecuteScalarAsync<Entity>(CancellationToken.None, query + param);                        // Noncompliant

        // All IAlterPoco and IAlterPocoAsync
        database.Update<Entity>(query + param);                                                            // Noncompliant
        database.Delete<Entity>(query + param);                                                            // Noncompliant
        database.UpdateAsync<Entity>(query + param);                                                       // Noncompliant
        database.UpdateAsync<Entity>(CancellationToken.None, query + param);                               // Noncompliant
        database.DeleteAsync<Entity>(query + param);                                                       // Noncompliant
        database.DeleteAsync<Entity>(CancellationToken.None, query + param);                               // Noncompliant
    }

    public void SqlType(string query, int param)
    {
        var s1 = new Sql(query + param);                                                                   // Noncompliant
        Sql s2 = new(query + param);                                                                       // Noncompliant
        Sql.Builder.Where(query + param);                                                                  // Noncompliant
    }
}

class Entity
{
    public int Id { get; set; }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/ExecutingSqlQueries.PetaPoco.vb
Imports System
Imports System.Linq
Imports System.Threading
Imports PetaPoco

Class Program
    Public Sub IExecuteMethods(ByVal execute As IExecute, ByVal query As String, ByVal param As String)
        execute.Execute(query)                                                                                      ' Compliant
        execute.Execute(query & param)                                                                              ' Noncompliant
        execute.ExecuteScalar(Of Entity)(query)                                                                     ' Compliant
        execute.ExecuteScalar(Of Entity)(query & param)                                                             ' Noncompliant
    End Sub

    Public Sub IDatabaseMethods(ByVal database As IDatabase, ByVal query As String, ByVal param As Integer, ByVal otherParam As Integer)
        database.Execute(query)                                                                                     ' Compliant
        database.Execute(query & param)                                                                             ' Noncompliant
        database.ExecuteScalar(Of Entity)(query)                                                                    ' Compliant
        database.ExecuteScalar(Of Entity)(query & param)                                                            ' Noncompliant
        database.Query(Of Entity)(query)                                                                            ' Compliant
        database.Query(Of Entity)(query & param)                                                                    ' Noncompliant
        database.Query(Of Entity)(query, param + otherParam)                                                        ' Compliant
        database.Query(Of Entity)({GetType(Entity)}, Nothing, query & param)                                        ' Noncompliant
        database.Query(Of Entity, Entity)(query & param)                                                            ' Noncompliant
        database.Query(Of Entity, Entity, Entity)(query & param)                                                    ' Noncompliant
        database.Query(Of Entity, Entity, Entity, Entity)(query & param)                                            ' Noncompliant
        database.Query(Of Entity, Entity, Entity, Entity, Entity)(query & param)                                    ' Noncompliant

        database.Fetch(Of Entity)(query)                                                                            ' Compliant
        database.Fetch(Of Entity)(query & param)                                                                    ' Noncompliant
        database.Fetch(Of Entity)(query & param)                                                                    ' Noncompliant
        database.Fetch(Of Entity)(2, 10, query & param)                                                             ' Noncompliant
        database.Fetch(Of Entity)(param + otherParam, 10, query)                                                    ' Compliant
        database.Page(Of Entity)(0, 0, query & param, New Object(-1) {}, query & param, New Object(-1) {})
        '                              ^^^^^^^^^^^^^
        '                                                                ^^^^^^^^^^^^^                                @-1
        database.Exists(Of Entity)(query & param)                                                                   ' Noncompliant
        database.Fetch(Of Entity, Entity)(query & param)                                                            ' Noncompliant
        database.First(Of Entity)(query & param)                                                                    ' Noncompliant
        database.FirstOrDefault(Of Entity)(query & param)                                                           ' Noncompliant
        database.Page(Of Entity)(0, 0, query & param)                                                               ' Noncompliant
        database.Page(Of Entity)(0, 0, query & param, Array.Empty(Of Object)(), "", Array.Empty(Of Object)())       ' Noncompliant
        database.Page(Of Entity)(0, 0, "", Array.Empty(Of Object)(), query & param, Array.Empty(Of Object)())       ' Noncompliant
        database.Query(Of Entity, Entity)(query & param)                                                            ' Noncompliant
        database.QueryMultiple(query & param)                                                                       ' Noncompliant
        database.SkipTake(Of Entity)(0, 0, query & param)                                                           ' Noncompliant
        database.Single(Of Entity)(query & param)                                                                   ' Noncompliant
        database.SingleOrDefault(Of Entity)(query & param)                                                          ' Noncompliant
        database.ExistsAsync(Of Entity)(query & param)                                                              ' Noncompliant
        database.FetchAsync(Of Entity)(query & param)                                                               ' Noncompliant
        database.FirstAsync(Of Entity)(query & param)                                                               ' Noncompliant
        database.FirstOrDefaultAsync(Of Entity)(query & param)                                                      ' Noncompliant
        database.PageAsync(Of Entity)(0, 0, query & param)                                                          ' Noncompliant
        database.PageAsync(Of Entity)(0, 0, query & param, Array.Empty(Of Object)(), "", Array.Empty(Of Object)())  ' Noncompliant
        database.PageAsync(Of Entity)(0, 0, "", Array.Empty(Of Object)(), query & param, Array.Empty(Of Object)())  ' Noncompliant
        database.QueryAsync(Of Entity)(query & param)                                                               ' Noncompliant
        database.SkipTakeAsync(Of Entity)(0, 0, query & param)                                                      ' Noncompliant
        database.SingleAsync(Of Entity)(query & param)                                                              ' Noncompliant
        database.SingleOrDefaultAsync(Of Entity)(query & param)                                                     ' Noncompliant
        database.Execute(query & param)                                                                             ' Noncompliant
        database.ExecuteAsync(CancellationToken.None, query & param)                                                ' Noncompliant
        database.ExecuteScalar(Of Entity)(query & param)                                                            ' Noncompliant
        database.ExecuteScalarAsync(Of Entity)(CancellationToken.None, query & param)                               ' Noncompliant
        database.Update(Of Entity)(query & param)                                                                   ' Noncompliant
        database.Delete(Of Entity)(query & param)                                                                   ' Noncompliant
        database.UpdateAsync(Of Entity)(query & param)                                                              ' Noncompliant
        database.UpdateAsync(Of Entity)(CancellationToken.None, query & param)                                      ' Noncompliant
        database.DeleteAsync(Of Entity)(query & param)                                                              ' Noncompliant
        database.DeleteAsync(Of Entity)(CancellationToken.None, query & param)                                      ' Noncompliant
    End Sub

    Public Sub SqlType(ByVal query As String, ByVal param As Integer)
        Dim s1 = New Sql(query & param)                                                                             ' Noncompliant
        Sql.Builder.Where(query & param)                                                                            ' Noncompliant
    End Sub
End Class

Class Entity
    Public Property Id As Integer
End Class
